File: LC133_CloneGraph/DFS/dfs.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    # DFS
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def cloneGraph(self, node: 'Node') -> 'Node':
        visited = {}
        nodeCopy = self.dfs(node, visited)
        return nodeCopy
    
    def dfs(self, node, visited):
        if node is None:
            return None
        
        if node.val in visited:
            return visited[node.val]

        copyNode = Node(node.val, [])
        visited[node.val] = copyNode
        
        for neighbor in node.neighbors:
            copyNeighber = self.dfs(neighbor, visited)
            copyNode.neighbors.append(copyNeighber)
        return copyNode

File: LC133_CloneGraph/DFS/dfs.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
using namespace std;

class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution {
    // DFS
    // Time Complexity: O(n)
    // Space COmplexity: O(n)
public:
    Node* cloneGraph(Node* node) {
        if(node == nullptr){
            return nullptr;
        }
        unordered_map<int, Node*> visited;
        return dfs(node, visited);
        
    }
    Node* dfs(Node* node, unordered_map<int, Node*>& visited){
        Node* nodeCopy;
        if(visited.count(node->val) == 0){
            nodeCopy = new Node(node->val);
            visited[node->val] = nodeCopy;
        }else{
            nodeCopy = visited[node->val];
            return nodeCopy;
        }

        for(Node* n : node->neighbors){
            if(n != nullptr){
                Node* nCopy = dfs(n, visited);
                nodeCopy->neighbors.push_back(nCopy);
            }
        }
        return nodeCopy;
    }

};